### PR TITLE
udev-util: ignore USB-C ports in power source mode

### LINKED ISSUE #21988: AC power check wrong for discrete GPUs with USB-C
**systemd version the issue has been seen with**
> $ systemctl --version
systemd 250 (250-4-arch)
+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified

**Used distribution**
 > Arch Linux

**Linux kernel version used** (`uname -a`)
 > `Linux kitsune.inthat.cloud 5.15.12-zen1-1-zen #1 ZEN SMP PREEMPT Wed, 29 Dec 2021 12:04:52 +0000 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Unexpected behaviour you saw**
 > Not on AC power reported for a system with only AC power (a desktop with a discrete GPU):
```
$ systemd-analyze condition ConditionACPower=yes
test.service: ConditionACPower=yes failed.
Conditions failed.
```

**Steps to reproduce the problem**
 > This is a system with a discrete GPU (AMD Radeon 6900 XT) that has a USB-C connector.  The USCI for that appears to reporting a power type of "USB" and offline (nothing connected).  Since there are no other power supplies, this is making systemd think the system is not on AC power. 

Basically: this system only has one entry in `/sys/class/power_supply/` but that entry is the power **output** from a discrete GPU (or at least it certainly can't run the system), which is reporting offline.  Since there is no other power available, the check `on_ac_power()` believes the system does not have AC power.   I believe the problem is c19a51bec40ae5e5073464e72411e7920d05d683, prior to that the check would have excluded the UCSI power.

**Additional program output to the terminal or log subsystem illustrating the issue**

```
$ ls /sys/class/power_supply
ucsi-source-psy-0-00081
```

```
$ grep . /sys/class/power_supply/ucsi-source-psy-0-00081/*
/sys/class/power_supply/ucsi-source-psy-0-00081/current_max:0
/sys/class/power_supply/ucsi-source-psy-0-00081/current_now:0
grep: /sys/class/power_supply/ucsi-source-psy-0-00081/device: Is a directory
grep: /sys/class/power_supply/ucsi-source-psy-0-00081/hwmon6: Is a directory
/sys/class/power_supply/ucsi-source-psy-0-00081/online:0
grep: /sys/class/power_supply/ucsi-source-psy-0-00081/power: Is a directory
grep: /sys/class/power_supply/ucsi-source-psy-0-00081/subsystem: Is a directory
/sys/class/power_supply/ucsi-source-psy-0-00081/type:USB
/sys/class/power_supply/ucsi-source-psy-0-00081/uevent:POWER_SUPPLY_NAME=ucsi-source-psy-0-00081
/sys/class/power_supply/ucsi-source-psy-0-00081/uevent:POWER_SUPPLY_TYPE=USB
/sys/class/power_supply/ucsi-source-psy-0-00081/uevent:POWER_SUPPLY_USB_TYPE=[C] PD PD_PPS
/sys/class/power_supply/ucsi-source-psy-0-00081/uevent:POWER_SUPPLY_ONLINE=0
/sys/class/power_supply/ucsi-source-psy-0-00081/uevent:POWER_SUPPLY_VOLTAGE_MIN=0
/sys/class/power_supply/ucsi-source-psy-0-00081/uevent:POWER_SUPPLY_VOLTAGE_MAX=0
/sys/class/power_supply/ucsi-source-psy-0-00081/uevent:POWER_SUPPLY_VOLTAGE_NOW=0
/sys/class/power_supply/ucsi-source-psy-0-00081/uevent:POWER_SUPPLY_CURRENT_MAX=0
/sys/class/power_supply/ucsi-source-psy-0-00081/uevent:POWER_SUPPLY_CURRENT_NOW=0
/sys/class/power_supply/ucsi-source-psy-0-00081/usb_type:[C] PD PD_PPS
/sys/class/power_supply/ucsi-source-psy-0-00081/voltage_max:0
/sys/class/power_supply/ucsi-source-psy-0-00081/voltage_min:0
/sys/class/power_supply/ucsi-source-psy-0-00081/voltage_now:0
grep: /sys/class/power_supply/ucsi-source-psy-0-00081/wakeup47: Is a directory
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #21988.

### CODE REVIEW

**Review by bluca — APPROVED** [review]

File: src/shared/udev-util.c [hunk 1]
```diff
@@ -584,59 +581,62 @@ int udev_queue_init(void) {
...
-        _cleanup_closedir_ DIR *d = NULL;
+        sd_device *d;
         int r;
 
-        d = opendir("/sys/class/power_supply");
-        if (!d)
-                return errno == ENOENT ? true : -errno;
+        r = sd_device_enumerator_new(&e);
+        if (r < 0)
+                return r;
```
Comment (keszybz) [inline src/shared/udev-util.c:651]: Hmm, what are the effects of this change if udevd is not running or not fully initialized?

Comment (yuwata) [reply src/shared/udev-util.c:651]: Hm? I thought that `sd_device_enumerator` does not require udevd, especially when `allow_uninitialized()` is called. Is it not correct?

**Comment by yuwata** [comment]
The reporter of the issue confirmed that this works fine. Upgrading the green label.

File: src/shared/udev-util.c (hunk 1 above)
Comment (keszybz) [reply src/shared/udev-util.c:651]: I think that was the idea when it was introduced in dee5e0b6c29ca22cbf67d28fbee195d70b0dde72. But we don't have any docs and I'm not sure that it still works like this… I think we should verify that it behaves as expected before merging this.

Comment (yuwata) [reply src/shared/udev-util.c:651]: After re-reading sd-device code, actually it does not depend on udevd if `sd_device_enumerator` with `sd_device_enumerator_allow_uninitialized()` called, and only `sd_device_get_sysattr_value()` and `sd_device_get_parent()` are called for enumerated devices.

```
$ sudo udevadm info --cleanup-db
$ ls /run/udev/*
/run/udev/control

/run/udev/data:

/run/udev/links:

/run/udev/rules.d:

/run/udev/static_node-tags:

/run/udev/tags:

/run/udev/watch:
$ SYSTEMD_LOG_LEVEL=debug build/systemd-ac-power
sd-device-enumerator: Scan all dirs
sd-device-enumerator: Scanning /sys/bus
sd-device-enumerator: Scanning /sys/class
AC: The power supply is currently online.
BAT0: The power supply is battery, ignoring.
port0: The USB type-C port is in power sink mode.
port1: The USB type-C port is in power sink mode.
ucsi-source-psy-USBC000:001: The USB type-C device has at least one port in power sink mode.
ucsi-source-psy-USBC000:001: The power supply is currently offline.
port0: The USB type-C port is in power sink mode.
port1: The USB type-C port is in power sink mode.
ucsi-source-psy-USBC000:002: The USB type-C device has at least one port in power sink mode.
ucsi-source-psy-USBC000:002: The power supply is currently online.
Found at least one online non-battery power supply, system is running on AC power.
```

Comment (yuwata) [reply src/shared/udev-util.c:651]: Note that `sd-device` depends on udevd when,
- sd_device_monitor is monitoring netlink message from udevd.
- sd_device_enumerator accepts only initialized (hence, processed by udevd) devices.
- sd_device object called with sd_device_get_property_value(), sd_device_get_is_initialized(), sd_device_get_usec_initialized(), sd_device_get_tag_first/next(), sd_device_get_current_tag_first/next(), sd_device_get_devlink_first/next(), sd_device_has_tag(), sd_device_has_current_tag(), and so on.

Comment (bluca) [reply src/shared/udev-util.c:651]: That's great info, can we get it documented somewhere please? Maybe docs/architecture.md?

Comment (yuwata) [reply src/shared/udev-util.c:651]: Sure, I will do that later.

File: src/shared/udev-util.c [hunk 2]
```diff
@@ -584,60 +581,68 @@ int udev_queue_init(void) {
...
+                log_device_debug(d, "The power supply is currently %s.", v > 0 ? "online" : "offline");
         }
 
+        if (found_online)
+                log_debug("Found at least one online non-battery power supply, system is running on AC power.");
+        else if (!found_offline)
+                log_debug("Found no offline non-battery power supply, assuming system is running on AC power.");
+        else
+                log_debug("All non-battery power supplies are offline, assuming system is running with battery.");
+
```
Comment (keszybz) [inline src/shared/udev-util.c:718]: Those logs are a nice addition.

**Review by keszybz — APPROVED** [review]

**Pull request merged**